packages/nominal-types/src/index.ts

```diff
@@ -0,0 +1,80 @@
+ * unverifiedName satisfies Brand<string, 'VerifiedName'>; // ERROR
+ * verifiedName satisfies Brand<'Bob', 'VerifiedName'>; // ERROR
+ * verifiedName satisfies Brand<'Alice', 'VerifiedName'>; // OK
+ * verifiedName satisfies Brand<string, 'VerifiedName'>; // OK
+ * ```
+ *
+ * @example Values can be tagged as compressed data
+ * ```ts
+ * const untaggedData = new Uint8Array([/* ... *\/]);
+ * const compressedData = untaggedData as CompressedData<typeof untaggedData, 'zstd'>;
+ *
+ * compressedData satisfies CompressedData<Uint8Array, 'zstd'>; // OK
+ * untaggedData satisfies CompressedData<Uint8Array, 'zstd'>; // ERROR
+ * ```
+ *
+ * @example Strings can be tagged as being encoded using a particular scheme
+ * ```ts
+ * const untaggedString = 'dv1ZAGvdsz5hHLwWXsVnM94hWf1pjbKVau1QVkaMJ92';
+ * const encodedString = untaggedString as EncodedString<typeof untaggedString, 'base58'>;
+ *
+ * encodedString satisfies EncodedString<'dv1ZAGvdsz5hHLwWXsVnM94hWf1pjbKVau1QVkaMJ92', 'base58'>; // OK
+ * encodedString satisfies EncodedString<string, 'base58'>; // OK
+ * encodedString satisfies EncodedString<string, 'base64'>; // ERROR
+ * untaggedString satisfies EncodedString<string, 'base58'>; // ERROR
+ * ```
+ *
+ * @example Brands, compression, and encodings can be combined
+ * ```ts
+ * const encodedCompressedString = 'abc' as Brand<
+ *     EncodedString<CompressedData<'abc', 'zstd'>, 'base64'>,
+ *     'Base64ZstdCompressedData'
+ * >;
+ *
+ * encodedCompressedString satisfies Brand<'abc', 'Base64ZstdCompressedData'>; // OK
+ * encodedCompressedString satisfies Brand<string, 'Base64ZstdCompressedData'>; // OK
+ * encodedCompressedString satisfies CompressedData<'abc', 'zstd'>; // OK
+ * encodedCompressedString satisfies CompressedData<string, 'zstd'>; // OK
+ * encodedCompressedString satisfies EncodedString<'abc', 'base64'>; // OK
+ * encodedCompressedString satisfies EncodedString<string, 'base64'>; // OK
+ * encodedCompressedString satisfies EncodedString<string, 'base58'>; // ERROR
+ * ```
+ *
+ * @example Custom nominal types
+ * ```ts
+ * type SweeteningSubstance = 'aspartame' | 'cane-sugar' | 'stevia';
+ * type Sweetener<T extends SweeteningSubstance> = NominalType<'sweetener', T>;
+ *
+ * declare function eat(food: Sweetener<Exclude<SweeteningSubstance, 'aspartame'>>): void;
+ *
+ * const dessert = 'ice-cream' as string & Sweetener<'aspartame'>;
+ * eat(dessert); // ERROR
+ * ```
+ *
+ * @packageDocumentation
+ */
+
+type CompressionFormat = 'zstd';
+type StringEncoding = 'base58' | 'base64';
+
+export type Brand<T, TBrandName extends string> = NominalType<'brand', TBrandName> & T;
+export type CompressedData<T, TCompression extends CompressionFormat> = NominalType<'compressionFormat', TCompression> &
+    T;
+export type EncodedString<T extends string, TEncoding extends StringEncoding> = NominalType<
+    'stringEncoding',
+    TEncoding
+> &
+    T;
+
+export type NominalType<TCategory extends `${string}`, TIdentifier extends string> = {
```

Comment:
nit: Is there any advantage of using

```
extends `${string}`
```

over `extends string`?

Reply:
Not anymore! This used to be ```__${string}``` and I didn't collapse it when I removed the `__`.